syntax = "proto3";

package order_queue;

service OrderQueueService {
    rpc Enqueue (EnqueueRequest) returns (EnqueueResponse); 
    rpc Dequeue (DequeueRequest) returns (DequeueResponse);
}

message EnqueueRequest {
    OrderUUID order_id = 1;
    OrderData order_data = 2;
}

message EnqueueResponse {}

message DequeueRequest {}

message DequeueResponse {
    OrderUUID order_id = 1;
    OrderData order_data = 2;
}

message OrderUUID {
    string value = 1;
}

message OrderData {
    User user = 1;
    string user_comment = 2;
    CreditCard credit_card = 3;
    Items items = 4;
    string discount_code = 5;
    string shipping_method = 6;
    string gift_message = 7;
    BillingAddress billing_address = 8;
    bool gift_wrapping = 9;
    bool terms_accepted = 10;
    NotificationPreference notification_preference = 11;

}

message User {
    string user_name = 1;
    string user_contact = 2;
}

message CreditCard {
    string card_number = 1;
    string expiration_date = 2;
    string cvv = 3;
}

message BillingAddress {
    string country = 1;
    string state = 2;
    string city = 3;
    string street = 4;
    string zip = 5;
}

message Items {
    repeated Item items = 1;
}

message Item {
    string name = 1;
    uint32 quantity = 2;
}

message NotificationPreference {
    repeated string preferences = 1;
}
